src/Select.js

```diff
@@ -569,30 +569,60 @@ var Select = React.createClass({
 			}
 		}
 
-		this.props.asyncOptions(input, (err, data) => {
-			if (err) throw err;
-			if (this.props.cacheAsyncResults) {
-				this._optionsCache[input] = data;
-			}
-			if (thisRequestId !== this._currentRequestId) {
-				return;
-			}
-			var filteredOptions = this.filterOptions(data.options);
-			var newState = {
-				options: data.options,
-				filteredOptions: filteredOptions,
-				focusedOption: this._getNewFocusedOption(filteredOptions)
-			};
-			for (var key in state) {
-				if (state.hasOwnProperty(key)) {
-					newState[key] = state[key];
+		if (typeof this.props.asyncOptions() !== 'undefined' &&
+			typeof this.props.asyncOptions().then === 'function') {
+			this.props.asyncOptions(input).then((data) => {
```

Comment:
But to be clear, you need to _also_ support the callback case here, so be sure to pass the `callback` argument to `asyncOptions` when you invoke it:

```
let result = this.props.asyncOptions(input, callback);
```

(haven't looked at how else you'd structure your code to allow this, I'm out and about at the moment but if you need more help I'll check back later)